cpp/src/arrow/dataset/filter.h

```diff
@@ -117,18 +86,95 @@ class ARROW_DS_EXPORT Expression {
 
   bool Equals(const std::shared_ptr<Expression>& other) const;
 
+  /// Overload for the common case of checking for equality to a specific scalar.
+  template <typename T, typename Enable = decltype(MakeScalar(std::declval<T>()))>
+  bool Equals(T&& t) const;
+
+  /// If true, this Expression is a ScalarExpression wrapping a null scalar.
+  bool IsNull() const;
+
   /// Validate this expression for execution against a schema. This will check that all
   /// reference fields are present (fields not in the schema will be replaced with null)
   /// and all subexpressions are executable. Returns the type to which this expression
   /// will evaluate.
   virtual Result<std::shared_ptr<DataType>> Validate(const Schema& schema) const = 0;
 
-  /// Return a simplified form of this expression given some known conditions.
-  /// For example, (a > 3).Assume(a == 5) == (true). This can be used to do less work
-  /// in ExpressionFilter when partition conditions guarantee some of this expression.
-  /// In the example above, *no* filtering need be done on record batches in the
-  /// partition since (a == 5).
-  virtual Result<std::shared_ptr<Expression>> Assume(const Expression& given) const {
+  Status Validate(const Schema& schema, std::shared_ptr<DataType>* out) const {
+    return Validate(schema).Value(out);
+  }
+
+  /// \brief Simplify to an equivalent Expression given assumed constraints on input.
```

Comment:
In the future, I think we should make Optimization/Simplification it's own class.